### PR TITLE
SLO Ratio Query 

### LINKED ISSUE #983: SLO: Add Ratio Query Type
- Add the Ratio Query type to the Grafana Terraform Provider, requires modifying the HCL Schema for both the resource and data source
- Update documentation and examples to use Ratio Query Type

### PR DESCRIPTION
**Pull request by elainevuong**
- adds the SLO Ratio Query type to the Grafana Terraform Provider modifying the HCL Schema for both the resource and data source
- updates documentation and examples to use Ratio Query Type

Closes #983

### CODE REVIEW

**Comment by elainevuong** [comment]
Quick video for demo purposes attached. 

https://github.com/grafana/terraform-provider-grafana/assets/73451363/a60acde9-e1a4-444f-88b8-eed6ec7f70b9

File: docs/data-sources/slos.md [hunk 1]
```diff
@@ -199,3 +200,13 @@ Read-Only:
 Read-Only:
 
 - `query` (String)
+
+
+<a id="nestedobjatt--slos--query--ratio"></a>
+### Nested Schema for `slos.query.ratio`
+
+Read-Only:
```
Comment (ellisda) [inline docs/data-sources/slos.md:208]: nit - Why does this say read-only? These are fields that user can edit in their terraform right?

I'm probably missing something, because a bunch of these sections include "read only" - forgive me if I'm unclear whether we expect others to consume this or the schema as documentation

File: docs/data-sources/slos.md [hunk 2]
```diff
@@ -199,3 +200,13 @@ Read-Only:
...
+
+- `group_by_labels` (List of String)
```
Comment (ellisda) [inline docs/data-sources/slos.md:210]: I this a good place to mention it's optional?

File: docs/resources/slo.md [hunk 3]
```diff
@@ -34,29 +34,29 @@ resource "grafana_slo" "test" {
...
-    value = "value"
+    key   = "slo"
+    value = "terraform"
   }
   alerting {
     fastburn {
       annotation {
         key   = "name"
-        value = "Critical - SLO Burn Rate Alert"
+        value = "SLO Burn Rate Very High"
```
Comment (ellisda) [inline docs/resources/slo.md:44]: nit - why change?  - does this match changes we made in alert rule generation or something?

File: docs/resources/slo.md [hunk 4]
```diff
@@ -159,17 +156,34 @@ Required:
 
 Required:
 
-- `freeform` (Block List, Min: 1, Max: 1) (see [below for nested schema](#nestedblock--query--freeform))
 - `type` (String) Query type must be one of: "freeform", "query", "ratio", or "threshold"
 
+Optional:
```
Comment (ellisda) [inline docs/resources/slo.md:161]: Instead of "Optional" you could say "must include exactly one of the following (based on type above)" ??

File: docs/resources/slo.md [hunk 5]
```diff
@@ -159,17 +156,34 @@ Required:
...
+### Nested Schema for `query.ratio`
+
+Required:
+
+- `success_metric` (String) Metric for success events (numerator)
+- `total_metric` (String) Metric for total events (denominator)
+
+Optional:
+
+- `group_by_labels` (List of String) Defines Group By Labels used for per-label alerting. Also appear as variables on SLO dashboards to enable filtering and aggregation
```
Comment (ellisda) [inline docs/resources/slo.md]: you could mention that it must adhere to prometheus label name schema (i.e. we'll reject creation if they try to use fancy chars)

see comment on https://github.com/grafana/slo/blob/558e3c81683d907b5f36accc81eaed086442d3d2/schemas/slo_api.cue#L136 for regex and possible wording

File: examples/resources/grafana_slo/resource_ratio.tf [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    value = "terraform"
+  }
+  alerting {
+    fastburn {
+      annotation {
+        key   = "name"
+        value = "SLO Burn Rate Very High"
+      }
+      annotation {
+        key   = "name"
```
Comment (ellisda) [inline examples/resources/grafana_slo/resource_ratio.tf]: shouldn't this be `key = "description", so that the generated alert has both a name, and a description annotation?

File: examples/resources/grafana_slo/resource_ratio.tf [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+      }
+    }
+
+    slowburn {
+      annotation {
+        key   = "name"
+        value = "SLO Burn Rate High"
+      }
+      annotation {
+        key   = "name"
```
Comment (ellisda) [inline examples/resources/grafana_slo/resource_ratio.tf]: `s/name/description/g` here too?

File: examples/resources/grafana_slo/resource_update.tf [hunk 8]
```diff
@@ -12,29 +12,28 @@ resource "grafana_slo" "update" {
...
     fastburn {
       annotation {
         key   = "name"
-        value = "Critical - SLO Burn Rate Alert"
+        value = "SLO Burn Rate Very High"
       }
-      label {
-        key   = "type"
-        value = "slo"
+      annotation {
```
Comment (ellisda) [inline examples/resources/grafana_slo/resource_update.tf:24]: I see that you're adding a second annotation (you want `description` here I think, not `name` again), but why did you remove the label as well?

Most alerts will have name and description annotations (helps with notifications), but we also allow labels. Why not show both?

File: examples/resources/grafana_slo/resource_update.tf [hunk 9]
```diff
@@ -12,29 +12,28 @@ resource "grafana_slo" "update" {
...
       annotation {
         key   = "name"
-        value = "Warning - SLO Burn Rate Alert"
+        value = "SLO Burn Rate High"
       }
-      label {
-        key   = "type"
-        value = "slo"
+      annotation {
+        key   = "name"
```
Comment (ellisda) [inline examples/resources/grafana_slo/resource_update.tf]: `s/name/description/g` here

File: internal/resources/slo/data_source_slo.go [hunk 10]
```diff
@@ -99,6 +100,22 @@ func unpackQuery(apiquery gapi.Query) []map[string]interface{} {
...
+			"total_metric":    apiquery.Ratio.TotalMetric.PrometheusMetric,
+			"group_by_labels": apiquery.Ratio.GroupByLabels,
+		}
+
+		ratio = append(ratio, body)
+		query["ratio"] = ratio
+
+		retQuery = append(retQuery, query)
+	}
+
```
Comment (ellisda) [inline internal/resources/slo/data_source_slo.go:118]: do you need error handling for case where `apiQuery.Type` didn't match either of the if blocks?

File: internal/resources/slo/data_source_slo.go [hunk 11]
```diff
@@ -99,6 +100,22 @@ func unpackQuery(apiquery gapi.Query) []map[string]interface{} {
 		retQuery = append(retQuery, query)
 	}
 
+	if apiquery.Type == gapi.QueryTypeRatio {
```
Comment (ellisda) [inline internal/resources/slo/data_source_slo.go:103]: nit - it'd be more idomatic in golang to use a `switch apiQuery.Type {` with two case statements (and a default if we know how to handle the case where it didn't match either). You could use `if {} else if {}` as well, but the switch is even clearer to reader that only one of the blocks is going to run

File: internal/resources/slo/resource_slo.go [hunk 12]
```diff
@@ -57,17 +57,44 @@ Resource manages Grafana SLOs.
...
 										Type:        schema.TypeString,
-										Optional:    true,
+										Required:    true,
 										Description: "Freeform Query Field",
 									},
 								},
 							},
 						},
+						"ratio": &schema.Schema{
+							Type:     schema.TypeList,
```
Comment (ellisda) [inline internal/resources/slo/resource_slo.go:72]: is this a list just because ratio contains more than one field? I'm a HCL noob, just curious. I almost thought it meant a list of ratio definitions.

File: internal/resources/slo/resource_slo.go [hunk 13]
```diff
@@ -338,6 +365,40 @@ func packQuery(query map[string]interface{}) (gapi.Query, error) {
 		return sloQuery, nil
 	}
 
+	if query["type"] == "ratio" {
+		ratioquery := query["ratio"].([]interface{})[0].(map[string]interface{})
```
Comment (ellisda) [inline internal/resources/slo/resource_slo.go:369]: is this type assertion safe? It would panic if `query["ratio"]` held something besides a list...

File: internal/resources/slo/resource_slo.go [hunk 14]
```diff
@@ -338,6 +365,40 @@ func packQuery(query map[string]interface{}) (gapi.Query, error) {
...
+		successMetric := ratioquery["success_metric"].(string)
+		totalMetric := ratioquery["total_metric"].(string)
+		groupByLabels := ratioquery["group_by_labels"].([]interface{})
+
+		var labels []string
+
+		for ind := range groupByLabels {
+			if groupByLabels[ind] == nil {
```
Comment (ellisda) [inline internal/resources/slo/resource_slo.go:377]: you might as well drop any empty-string labels (i.e. length < 1). The backend will do the same, but only cause the FE had a similar bug. We might in the future,  start rejecting SLOs because empty-string isn't a valid prometheus label name

Comment (ellisda) [reply internal/resources/slo/resource_slo.go:377]: not a big deal, but would be better not to send emptystrings in the list

File: internal/resources/slo/resource_slo.go [hunk 15]
```diff
@@ -57,17 +57,44 @@ Resource manages Grafana SLOs.
...
+										Required:    true,
+									},
+									"total_metric": &schema.Schema{
+										Type:        schema.TypeString,
+										Description: `Metric for total events (denominator)`,
+										Required:    true,
+									},
+									"group_by_labels": &schema.Schema{
+										Type:        schema.TypeList,
+										Description: `Defines Group By Labels used for per-label alerting. Also appear as variables on SLO dashboards to enable filtering and aggregation `,
```
Comment (ellisda) [inline internal/resources/slo/resource_slo.go]: if this is the place end users will find documenation, then we should mention here that GroupByLabels are optional, but need to confirm to prometheus label name syntax 

https://github.com/grafana/slo/blob/558e3c81683d907b5f36accc81eaed086442d3d2/schemas/slo_api.cue#L135

File: internal/resources/slo/resource_slo.go [hunk 16]
```diff
@@ -57,17 +57,44 @@ Resource manages Grafana SLOs.
...
 						},
+						"ratio": &schema.Schema{
+							Type:     schema.TypeList,
+							MaxItems: 1,
+							Optional: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"success_metric": &schema.Schema{
+										Type:        schema.TypeString,
+										Description: `Metric for success events (numerator)`,
```
Comment (ellisda) [inline internal/resources/slo/resource_slo.go]: nit - you might say "counter metric" to help clarify that we don't want them to add "rate", etc. They'll figure it out from example, but...

**Review by ellisda — APPROVED** [review]
LGTM. A made a couple of nit comments for you to consider, but no changes really needed.

I was hoping that schema (and/or the .md?? -- which will end users see as docs??) could follow the comments we've made in https://github.com/grafana/slo/blob/558e3c81683d907b5f36accc81eaed086442d3d2/schemas/slo_api.cue#L41... but :cry:  we don't have many comments there.

and while it should match https://github.com/grafana/slo/blob/4a46b5ccc4a7ced0894e2f1d43f7a7d4100d62f9/openapi/slo.yaml#L105 you're ahead of that, too :- )

File: docs/data-sources/slos.md (hunk 1 above)
Comment (elainevuong) [reply docs/data-sources/slos.md:208]: So this is the README for the SLOs Datasource - basically Terraform has two types of resources:
resources and datasources.
Datasources allow Terraform to read objects in - so that's why all the fields within this datasource markdown are read-only fields.

File: docs/resources/slo.md (hunk 3 above)
Comment (elainevuong) [reply docs/resources/slo.md:44]: Yep it does - I modified the key-value pairs within our annotations to match the default values that come out when the user clicks "Add Alert Rules" from the UI, just to make it crystal clear. Also if users decide to start with our documentation as a place to make SLOs with Terraform, it should create a SLO in a very similar format as if they had used the UI.

File: internal/resources/slo/resource_slo.go (hunk 12 above)
Comment (elainevuong) [reply internal/resources/slo/resource_slo.go:72]: Yeah it's a TypeList because I wanted the nesting, but the list has a restriction of being only a length of one (hence the line `MaxItems: 1`. And then that singular item is a struct consisting of `success_metric`, `total_metric`, and `group_by_labels`

File: internal/resources/slo/resource_slo.go (hunk 13 above)
Comment (elainevuong) [reply internal/resources/slo/resource_slo.go:369]: We are expecting value within the `query["ratio"]` field to be in a list format, I think this is fine.

File: internal/resources/slo/data_source_slo.go (hunk 10 above)
Comment (elainevuong) [reply internal/resources/slo/data_source_slo.go:118]: This is on the READ Path from our SLO API back to Terraform (not from the User Input into Terraform) so we should always have the apiQuery type being either Ratio or Freeform. 

The user also gets a validation error if they input anything other than freeform or ratio within the type on the terraform state file as well.
